### DISCUSSION TITLE
Bot Senses and Memory Update

### DISCUSSION THREAD

**Discussion by T-Rizzle12** [T-Rizzle12/Custom-Gmod-Bot#29] [Announcements]
There were a few things I wanted to change about the bot. I wanted to give the bot a LOS, “Line of Sight,” a Hearing Interface and a Memory Interface. While I was creating said interfaces I also optimized and improved the code in the addon. As well as fixing some bugs.

For the Vision interface, it will check if the enemy is visible and in the bot's FOV, "Field Of View". The bot’s LOS will be affected by fog.

For the Hearing interface this allows the bot to "hear" enemies that are behind it. I use this to make the bot hear enemies approaching from behind the bot. Mostly to help with the fact the bot can’t see behind it anymore.

For the Memory Interface, this is where the bot will keep track of enemies it has seen before and check if they are still visible, “it will use the old visibility checks when checking LOS.”

Change Log:
- Added bot Vision Interface, the bot can no longer see behind it
- Added bot Hearing Interface, the bot can hear enemys nearby it and through walls
- Made the bot react to damage taken from an enemy
- The bot will prioritize enemies in its LOS unless an enemy on the bots known enemy list is closer than 200 HUs
- Added new Function IsHiddenByFog
- Changed current client file to a universal client file
- Increased the time between FireWeaponInterval, "when the bot presses its attack key"
- Made the bots cooldown for pressing buttons when using its medkit to use FireWeaponInterval
- Fixed a bug where the bot would use shotguns at long range and snipers at close range
- Added LookTarget, LookTargetTime, and LookTargetPriority
- LookTarget is the position the bot is trying to look at
- LookTargetTime is how long the bot should look at said position
- LookTargetPriority is how important the current position is. This allows me to set climbing up ladders to be more important than aiming at enemies
- Major optimization to weapon selection system
- Optimization to think function, it is now split into 2 functions, upkeep is run often and has important functions while update has functions that don’t need to be run every second
- Major optimization to distance checks, they now use lengthSqr instead on length
- Partial fix to the bot randomly jumping on stairs
- Changed medkit use interval to 0.5 seconds
- Moved the flashlight check to update think, since it effects every bot
- Moved head detection trace to upkeep function
- Fixed a potential bug where bot think timers would run on real players and/or bots that are not apart of my addon
- Added new function TBotBlind this function will be used to make the bot blind for the specified amount of time
- Limited how often the bot can swap it’s active weapon
- Changed how the bot presses its buttons, this new method allows me to tell the bot to push buttons without having it be in the start command hook
- Fixed the new think function :)
- The bot will no longer spam its attack button when an enemy is within melee distance, it caused the bot to accidentally attack teammates
- Replaced my think system with Valve’s bot think system, "this optimized code by 40%"
- Fixed a bug in PointWithinCursor sometimes causing the bot to press its IN_ATTACK when facing the opposite direction from its current enemy
- Optimized PointWithinCursor by doing the trace check last, "The trace check is the most expensive part, if the function will return false with the cheaper methods its should"
- Improved ladder path creation allowing me to remove the hack I used for it
- Removed IsValid checks from ShouldDropDown and ShouldJump since it caused them to always return false
- Changed how the bot looks at the path it’s following
- Partially changed how waypoints are created
- Changed Owner to TBotOwner, since Owner is used internally by the engine
- Added IsDoorOpen, the bot will no longer open doors if it is already open
- Fixed the bot only climbing the highest ladder if the bot has to climb up two or more ladders that share a nav_area to get on and off said ladders
- Made some fixes to the hybrid version of PathFollower
- Added functions to find hiding spots, “currently not used”
- Changed how the bot presses its movement keys, "WASD keys"
- Added new function GetDefaultFOV, "returns the default FOV of the bot"

<hr /><em>This discussion was created from the release <a href='https://github.com/T-Rizzle12/Custom-Gmod-Bot/releases/tag/V1.0.6'>Bot Senses and Memory Update</a>.</em>

**Comment by strubium** [comment]
Oh Yeah!
